pycromanager/test/conftest.py

```diff
@@ -0,0 +1,118 @@
+    url = "https://download.micro-manager.org" + m.group(1)
+
+    # download
+    print(f"\nDownloading Micro-manager nightly build: {url.split('/')[-1]}")
+    mm_installer = os.path.join(os.getcwd(), 'mm_nightly_build.exe')
+    if not os.path.exists(mm_installer):
+        wget.download(url, out=mm_installer)
+
+    yield mm_installer
+
+    # cleanup
+    os.remove(mm_installer)
+
+
+@pytest.fixture(scope="session")
+def install_mm(download_mm_nightly):
+    mm_installer = download_mm_nightly
+    mm_install_dir = os.path.join(os.path.expanduser('~'), "Micro-Manager-nightly")
+    mm_install_log_path = os.path.join(os.path.dirname(mm_installer), "mm_install.log")
+
+    # check that java dependencies are present
+    if os.path.isdir('java'):
+        java_path = os.path.abspath('java')
+    # in case cwd is '/pycromanager/test'
+    elif os.path.isdir('../../java'):
+        java_path = os.path.abspath('../../java')
+    else:
+        raise RuntimeError('Could not find pycro-manager/java path')
+
+    if not os.path.isdir(os.path.join(java_path, 'target')) \
+            or not os.path.isdir(os.path.join(java_path, 'target/dependency')):
+        raise FileNotFoundError('Please build Java dependencies before running pytest.')
+
+    # remove install dir if it exists, better to remove it at cleanup instead
+    if os.path.exists(mm_install_dir):
+        print(f'Removing previous Micro-manager installation at: {mm_install_dir}')
+        shutil.rmtree(mm_install_dir)
+    os.mkdir(mm_install_dir)
+
+    print(f'Installing Micro-manager nightly build at: {mm_install_dir}')
+    cmd = f"{mm_installer} /SP /VERYSILENT /SUPRESSMSGBOXES /CURRENTUSER /DIR={mm_install_dir} /LOG={mm_install_log_path}"
+    subprocess.run(cmd, shell=True)
+
+    # update pycromanager jar files with those newly compiled
+    replace_jars(os.path.join(java_path, 'target'), os.path.join(mm_install_dir, 'plugins', 'Micro-Manager'),
+                 ['PycroManagerJava'])
+    replace_jars(os.path.join(java_path, 'target/dependency'), os.path.join(mm_install_dir, 'plugins', 'Micro-Manager'),
+                 ['AcqEngJ', 'NDTiffStorage', 'NDViewer'])
+
+    yield mm_install_dir
+
+    # cleanup
+    os.remove(mm_install_log_path)
+    # fails, because MM is still running, I think
+    # shutil.rmtree(mm_install_dir)
+
+
+@pytest.fixture(scope="session")
+def setup_data_folder():
+    data_folder_path = os.path.join(os.getcwd(), 'temp_data')
+    if not os.path.isdir(data_folder_path):
+        os.mkdir(data_folder_path)
+
+    yield data_folder_path
+
+    shutil.rmtree(data_folder_path)
+
+
+@pytest.fixture(scope="session")
+def launch_mm_headless(install_mm):
+    mm_install_dir = install_mm
+    config_file = os.path.join(mm_install_dir, 'MMConfig_demo.cfg')
+
+    print('Launching Micro-manager in headless mode.')
+    start_headless(mm_install_dir, config_file)
+
+    # yield None
+    #
+    # cleanup()
```

Comment:
@henrypinkard can I get your help here? I need a way to shut down MM running in headless mode, but calling `acq_util.cleanup` is not doing it.

Reply:
Hmm weird. It seems to work for me (on mac). What OS are you on?

Reply:
You're right, `cleanup()` does shut down MM, I just needed to give it a few seconds (`time.sleep(5)`) before `shutil.rmtree(mm_install_dir)` is called by the next function. However, these the `cleanup()` function (I think) does cause `pytest` to hang, bost locally and in Github Actions

Reply:
Hmm any ideas why? I'm happy to play around with it if you can point to a minimal example of this

Reply:
Yeah, I can't figure it out...

I just modified the setup for the tests such that MM is installed only if it's not found in `~/Micro-Manager-nightly`. You should be able to run the test locally on your Mac if you place a working MM installation there. 

You'll find that if you uncomment these three lines pytest will hang up at the very end. The reason I want to properly shut down this headless instance is that I want to then test starting another one on a different port, or test starting two headless instances simultaneously.

Reply:
I just tried this on Mac and it didn't seem to cause problems. Seems like it must be a windows thing, which would make sense because there are differences in how Unix vs windows processes work. I can try also doing it on windows to see if I can get to the bottom of it

Reply:
Tried on windows too. All the tests pass and cleanup gets called as expected and it does indeed shutdown the Java process. I did manually install the latest nightly build. Maybe try that?

Reply:
I still get the same behavior on Windows - the tests complete, but then pytests hangs up before fully exiting. This seems to be a problem in Github Actions too - the tests stall for 6 hours at which point they can cancelled: https://github.com/czbiohub/pycro-manager/actions/runs/3753840267.

For now, I've commented out the call to `cleanup()` again. Let's fix this in another PR